url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/spec/args/StdArgList.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Scripty Programming Language
+ * Copyright (C) 2010-2012 Bruno Ranschaert, S.D.I.-Consulting BVBA
+ * http://www.sdi-consulting.be
+ * mailto://info@sdi-consulting.be
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package com.sdicons.scripty.spec.args;
+
+import com.sdicons.scripty.parser.IContext;
+import com.sdicons.scripty.parser.Pair;
+
+/**
+ * There are 3 types of parameters in a standard argument list.
+ * 1. Fixed and required arguments. Each with its own type.
+ * 2. Optional. Each argument has its own type. These cannot be of type Pair, because this might conflict with the named arguments.
+ *    The optional parameters can have a default value which will be used when the argument is not present.
+ * 3. Named (optional or required). These are pairs at the end of the command line.
+ *    The named parameters can have a default value.
+ */
+public class StdArgList 
+implements IArgList
+{
+    public static final IArgList NOARG = new StdArgList(new FixedArg[]{}, new OptionalArg[]{}, new NamedArg[]{});
+    
+    private FixedArg req[];
+    private OptionalArg[] opt;
+    private NamedArg[] named;
+    
+    public StdArgList(FixedArg aReq[], OptionalArg[] aOpt, NamedArg[] aNamed)
+    {
+        req = aReq;
+        opt = aOpt;
+        named = aNamed;
+    }
+    
+    /**
+     * The ars are expected to have the form ("CMD", arg-1, arg-2, ... arg-n).
+     */
+    public Object[] guard(Object[] aArgs, IContext aCtx) 
+    throws ArgSpecException
+    {
+        // Create a new argument list where we will accumulate the
+        // converted results.
+        Object[] lNewArgs = new Object[1 + req.length + opt.length + named.length];
+        // Copy the command name to the new argument list, the structure will remain the same.
+        lNewArgs[0] = aArgs[0];
+        
+        // Check the fixed.
+        if(aArgs.length - 1 < req.length)
+            throw new ArgSpecException(String.format("Too few arguments. Expected at least %d arguments but received %d.", req.length, aArgs.length - 1));
+        
+        // We skip the command name.
+        int lArgIdx = 1;
+        for(int i = 0; i < req.length; i++) 
+        {
+            lNewArgs[lArgIdx] = req[i].guard(aArgs, lArgIdx++, aCtx);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;aCtx&#39; (index: 1 | type: IContext). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;aArgs&#39;. The same parameter type is validated here for example: url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java:60. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).